arch/x86/core/intel64/locore.S

```diff
@@ -21,8 +22,65 @@
 	pop %rax
 .endm
 
+/*
+ * Definitions/macros for enabling paging
+ */
+
+/* Long mode, no-execute, syscall */
+#define EFER_BITS (X86_EFER_MSR_LME | X86_EFER_MSR_NXE | X86_EFER_MSR_SCE)
+
+/* Paging, write-protect */
+#define CR0_BITS (CR0_PG | CR0_WP)
+
+/* PAE, SSE */
+#define CR4_BITS (CR4_PAE | CR4_OSFXSR)
+
+.macro set_efer
+	movl $X86_EFER_MSR, %ecx
+	rdmsr
+	orl $EFER_BITS, %eax
+	wrmsr
+.endm
+
+.macro install_pagetables_32
+	movl %cr4, %eax
+	orl $CR4_BITS, %eax
+	movl %eax, %cr4
+	clts
+
+	/* Page tables created at build time by gen_mmu.py */
+	movl $Z_MEM_PHYS_ADDR(z_x86_kernel_ptables), %eax
+	movl %eax, %cr3
+
+	set_efer
+
+	movl %cr0, %eax
+	orl $CR0_BITS, %eax
+	movl %eax, %cr0
+.endm
+
+.macro install_pagetables_64
+	/* Here, we are already in long mode with paging enabled and
+	 * just need to switch to our own page tables, but let's be
+	 * paranoid and ensure CR4, CR0, and EFER_MSR are set up
+	 * exactly how we expect. Logic is the same as install_pagetables_32
+	 */
+	movq %cr4, %rax
+	orq $CR4_BITS, %rax
+	movq %rax, %cr4
+	clts
+
+	movq $Z_MEM_PHYS_ADDR(z_x86_kernel_ptables), %rax
+	movq %rax, %cr3
+
+	set_efer
+
+	movq %cr0, %rax
+	orl $CR0_BITS, %eax
```

Comment:
`%rax`?

Reply:
let me double check this, I vaguely remember I did this intentionally (64-bit instructions can be weird with respect to immediate values, and all the CR0_BITS are in low 32-bits) but if I did there should be a comment explaining why.

Reply:
Yeah this is some fussiness from the assembler.

If I change to:

```
	orq $CR0_BITS, %rax
```

Then the assembler complains:

```
/home/apboie/projects/zephyr/zephyr/arch/x86/core/intel64/locore.S: Assembler messages:
/home/apboie/projects/zephyr/zephyr/arch/x86/core/intel64/locore.S:239: Error: operand type mismatch for `or'
```

This is rooted in CR0_PG being bit 31.
Added a comment.

Reply:
by the way setting `cr0` here is me being pedantic; 64-bit mode requires paging already be turned on but this is to make extra-sure the WP bit is set.